KEP 3140: TimeZone support in CronJob

soltysh opened the pull request:
- One-line PR description: TimeZone support in CronJob
- Issue link: https://github.com/kubernetes/enhancements/issues/3140

/assign @ravisantoshgudimetla @janetkuo 
/cc @iterion

/sig apps

soltysh commented:
/assign @deads2k

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+
+### Goals
+
+- Add the field `.spec.timeZone` which allows specifying a valid TimeZone
```
Is this a name or an offset from UTC?

soltysh replied:
name - updated

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
```
Whichever sense of the `.spec.timeZone` field we use (name or offset from UTC), let's list the other approach as an alternative.

soltysh replied:
Done.

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/kep.yaml:
```diff
@@ -0,0 +1,44 @@
+      - kube-controller-manager
+disable-supported: true
+
+# The following PRR answers are required at beta release
+metrics:
+  - my_feature_metric
```
Maybe comment this out?

soltysh replied:
Done.

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
````diff
@@ -0,0 +1,504 @@
+}
+```
+
+The value provided in `TimeZone` field will be validated against the embedded
+golang timezone database.
````
Does this mean that changes to time zone databases need to be applied to the API server and kube-controller-manager in a particular order?

soltysh replied:
No need for manual action, just follow the general guideline for updating the cluster: https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+
+### CronJob API
+
+The `CronJobSpec` structure is expanded with new `TimeZone` field which allows
+specifying the name of the time zone to be used. Missing or empty value of the
+field indicates the current behavior, which relies on the time zone of the
+kube-controller-manager process.
```
```suggestion
The `.spec` for a CronJob is expanded with new `timeZone` field which allows
specifying the name of the time zone to be used. Missing or empty value of the
field indicates the current behavior, which relies on the time zone of the
kube-controller-manager process.

In the API code, that looks like:
```

1. Does the kube-controller-manager populate the `.spec.timeZone` once it detects a CronJob with no timezone set?
2. Should it?

janetkuo commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+
+Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
+will take the field into account when scheduling the next Job run. In case the
+field is not specified or is empty the controller will maitain the current
```
```suggestion
field is not specified or is empty the controller will maintain the current
```

janetkuo commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+time zone database up-to-date.
+
+- Malicious user can create multiple CronJobs with different time zone which can
+actually trigger Jobs at the exact same time
```
Can't malicious user trigger Jobs at the same time without time zone?

soltysh replied:
It's easier to notice if all jobs are expressed w/o timezone, with timezone it's slightly harder.

janetkuo commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
````diff
@@ -0,0 +1,504 @@
+
+```golang
+
+type CronJobSpec struct {
+
+	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
+	Schedule string
+
+  // Time zone for the above schedule
+  TimeZone *string
+
+}
````
```suggestion
type CronJobSpec struct {

	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
	Schedule string

	// Time zone for the above schedule
	TimeZone *string

}
```

soltysh replied:
Fixed.

atiratree commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,504 @@
+on the time zone of the kube-controller-manager process which is hard for a
+regular user to figure out. Exposing an explicit field for setting a time zone
+will allow CronJob authors to simplify the user experience when creating CronJobs.
+
```
Would it make sense to also introduce a way to set a global default TZ?

sftim replied:
Ah, I had an idea about that.
_In my view_ that is best left for an external controller or for an admission webhook. However, the “external” mechanism could still be official Kubernetes code.

soltysh replied:
Yes, no global setting, we don't have it anywhere, and I'd leave it to cluster admin to implement, re-use such setting.

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+
+## Proposal
+
+Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
```
We could make the field a bit more self-documenting
```suggestion
Add the field `.spec.timeZoneName` to the CronJob resource. The cronjob controller
```

soltysh replied:
I think using time zone is in most cases sufficient.

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+
+CronJob controller will use non-nil, non-empty value from `TimeZone` field when
+parsing the schedule and during scheduling the next run time. Additionally the
+time zone will be reflected in the `.status.lastSuccessfulTime` and `.status.lastScheduleTime`.
```
We could also record the applied offset from UTC (from the last invocation) in `.status`. Maybe after an alpha though?

soltysh replied:
`lastSuccessfulTime` is of `metav1.Time` type which embeds `time.Time` which already has time zone information embeded in it, so I'm planning to re-use that.

sftim commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```
I suppose the other other alternative is timezone details in the cron time spec (the approach that got merged by accident).

soltysh replied:
Rarely, we re-use the current fields, that case you're talking about is more a feature creep not intentional on our side, b/c it exposes implementation details, that were not covered in the original design .

sftim replied:
The inline timezone spec - as mentioned in https://web.archive.org/web/20211007074332/https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/ - is not the design we want. I get that. We should nonetheless list it as a ruled-out alternative.

sftim replied:
I can PR that in after this merges.

ravisantoshgudimetla reviewed (approved):
/lgtm

Thank you for working on this @soltysh. This will be really helpful for the community.

deads2k commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+
+When downgrading from a release with this feature, to a release without `TimeZone`
+there are a few cases:
+  1. If TimeZone feature gate was enabled and user specified a TimeZone, the newly
```
How will you test to be sure the cronjob controller correctly handles going from timezone aware to non-timezone aware?

soltysh replied:
Manual testing of a cronjob with time zone specified, I can't think of a better approach.

soltysh replied:
I've mentioned that in the test plan, that we want to enabled with the feature on and off to ensure the controller behaves as expected. As much as possible those tests will be automated in integration, but not everything is possible there.

deads2k commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, the feature can be disabled. The outcome of disabling will depend if the
```
CronJob controller has had difficulties in the past.  How will you test the enable, disable, enable flow?

soltysh replied:
Similarly as with the previous, a manual testing will be performed to ensure this works as expected.

deads2k replied:
unit tests from line 272 will work.

deads2k commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
```
An upgrade flow can be vulnerable to the enable, disable, enable if you have a lease that is acquired by a new KCM, then an old KCM, then a new KCM.  The unit and integration tests need to be robust.

soltysh replied:
ack, added that here

deads2k commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,509 @@
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
```
While not required for alpha, can you think of what metrics you'll watch?  If you're able to fill this in now, people running the alpha will know what information to look for.

dims commented:
@soltysh 
- can you please add examples of "timezone name"? 
- https://pkg.go.dev/time/tzdata mentions that the binary size is going to be larger when `time/tzdata` is imported, do we want to mention that somewhere? ( looking at https://pkg.go.dev/time/tzdata )

deads2k commented:
PRR is complete for alpha.

/approve

soltysh commented:
@dims
> * can you please add examples of "timezone name"?

Added link to https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

> * https://pkg.go.dev/time/tzdata mentions that the binary size is going to be larger when `time/tzdata` is imported, do we want to mention that somewhere? ( looking at https://pkg.go.dev/time/tzdata )

Added that information, even though the size is negligible since according to my test it's less than 500kB. 

Both are in the last commit.

atiratree commented:
/lgtm

dims commented:
thanks @soltysh !

BenTheElder commented on keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md:
```diff
@@ -0,0 +1,517 @@
+at the time was that introducing such functionality would require cluster operator
+to manually include TimeZone database since golang did not have one.
+Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
+be embedded in the binary thus removing the requirement for external database.
```
I think we should explicitly call out in the KEP intent to embed this package as the approach for keeping tzdata up to date?

(Or else, if we have another approach planned, we should outline it?)

sftim replied:
Hmm. If there is a timezone change (eg a new national law), does that trigger a patch release?

sftim replied:
The approach I initially put forward was to have Kubernetes itself only aware of offsets from UTC.

BenTheElder replied:
The implementing PR appears to be taking the `import _ "time/tzdata"` approach, operators would have the option to patch their deployments with images or mounts containing a non-embedded on-disk tzdata and go will prefer that.

I'm also not sure that the embedded copy will be used at all in official releases / it's not clear how helpful importing this is, all of our images should be shipping with tzdata on-disk in the image. will x-ref the PR.

soltysh replied:
I've already responded to that other comment, but I'll leave the justification also here. The intent why we picked this approach, as a safe net was mentioned here in the motivation section. I'll make sure to add some more details during beta reviews to provide more details around expectations and assumptions for the embedded vs the on-system tz database.
